include/libtorrent/announce_entry.hpp

```diff
@@ -62,13 +91,7 @@ namespace libtorrent
 		// trackerid is sent).
 		std::string trackerid;
 
-		// if this tracker has returned an error or warning message
-		// that message is stored here
-		std::string message;
-
-		// if this tracker failed the last time it was contacted
-		// this error code specifies what error occurred
-		error_code last_error;
+		std::vector<announce_endpoint> endpoints;
```

Comment:
I didn't realize the list was (tracker, listen_socket) pairs, I thought it was just a list of each endpoint a tracker hostname resolved to. I think you make good points to keep these as a list on the tracker then.
regarding the ``announce_to_all_tiers``, I don't believe it's possible to deprecate that. My impression is that people creating torrents don't understand the announce rules defined by the multi-tracker BEP enough to create reasonable torrents, and since torrents won't work well without it, clients default to announcing to all in parallel.
(unless of course you mean to deprecate the original behavior of announcing to trackers one tier at a time. I would be a bit worried about people actually relying on that behavior, since it's quite useful and pretty well specified).

That said, the existing logic in libtorrent to iterate over trackers and decide which one to announce to is a mess (and quite error prone), and I apologize if you've had to read that :/ I've made a few attempts to clean it up a bit.